File: examples/example_basic.py
"""
Example 1: Basic eval suite — no external APIs required.

Demonstrates:
  - Defining test cases inline
  - Multiple scoring strategies (exact, fuzzy, contains, custom, json)
  - Running the harness synchronously
  - Saving reports

Run:
    cd examples
    pip install -e ../python
    python example_basic.py
"""

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "python"))

from evalforge import EvalHarness, TestCase
from evalforge.scorer import (
    exact_match,
    fuzzy_match,
    contains_match,
    json_match,
    custom_scorer,
)


# ---------------------------------------------------------------------------
# Define the "agent" we're evaluating
# This is a simple mock; in real usage this would call your LLM or system.
# ---------------------------------------------------------------------------

def echo_agent(input_data: str) -> str:
    """A trivial agent that echoes back a response based on input."""
    responses = {
        "What is 2 + 2?": "4",
        "What is the capital of France?": "The capital of France is Paris.",
        "Tell me a joke": "Why did the chicken cross the road? To get to the other side!",
        "Say hello": "Hello! How can I help you today?",
        "Return JSON": '{"name": "Alice", "age": 30, "city": "NYC"}',
    }
    return responses.get(input_data, "I don't know.")


# ---------------------------------------------------------------------------
# Define the test suite
# ---------------------------------------------------------------------------

def make_test_suite() -> list[TestCase]:
    def score_contains_number(expected, actual) -> float:
        """Custom scorer: check if the actual output contains any digit."""
        return 1.0 if any(c.isdigit() for c in str(actual)) else 0.0

    return [
        TestCase(
            id="math-basic",
            description="Simple arithmetic",
            input="What is 2 + 2?",
            expected_output="4",
            scoring=exact_match(),
            tags=["math", "basic"],
        ),
        TestCase(
            id="geography-capital",
            description="Capital city lookup",
            input="What is the capital of France?",
            expected_output="Paris",
            scoring=contains_match(),
            tags=["geography"],
        ),
        TestCase(
            id="greeting-fuzzy",
            description="Greeting should be warm",
            input="Say hello",
            expected_output="Hello! How can I help?",
            scoring=fuzzy_match(threshold=0.7),
            tags=["greeting"],
        ),
        TestCase(
            id="json-output",
            description="Agent returns valid JSON with expected fields",
            input="Return JSON",
            expected_output={"name": "Alice", "age": 30},
            scoring=json_match(ignore_keys=["city"], threshold=1.0),
            tags=["structured"],
        ),
        TestCase(
            id="custom-number-check",
            description="Response must contain a number",
            input="What is 2 + 2?",
            expected_output="4",
            scoring=custom_scorer(score_contains_number, threshold=1.0),
            tags=["math", "custom"],
        ),
        TestCase(
            id="unknown-input",
            description="Agent gracefully handles unknown queries",
            input="xyzzy-unknown-prompt",
            expected_output="I don't know.",
            scoring=exact_match(),
            tags=["edge-case"],
        ),
    ]


# ---------------------------------------------------------------------------
# Run
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    harness = EvalHarness(
        agent=echo_agent,
        suite_name="basic-echo-suite",
        default_timeout=5.0,
        verbose=True,
    )

    harness.add_many(make_test_suite())

    result = harness.run(
        report_json="reports/basic_report.json",
        report_html="reports/basic_report.html",
    )

    # Programmatic access
    print(f"\nProgrammatic access: pass_rate={result.pass_rate:.1%}, avg_score={result.avg_score:.3f}")

    # Exit non-zero if any tests failed
    sys.exit(0 if result.pass_rate == 1.0 else 1)

File: typescript/examples/example_basic.ts
/**
 * Example 1: Basic eval suite (TypeScript) — no external APIs required.
 *
 * Run:
 *   cd typescript
 *   npm install
 *   npx ts-node --esm examples/example_basic.ts
 */

import { EvalHarness, TestCase } from "../dist/index.js";
import {
  exactMatch,
  fuzzyMatch,
  containsMatch,
  jsonMatch,
  customScorer,
} from "../dist/scorer.js";

// ---------------------------------------------------------------------------
// Mock agent
// ---------------------------------------------------------------------------

const RESPONSES: Record<string, string> = {
  "What is 2 + 2?": "4",
  "What is the capital of France?": "The capital of France is Paris.",
  "Tell me a joke": "Why did the chicken cross the road? To get to the other side!",
  "Say hello": "Hello! How can I help you today?",
  "Return JSON": '{"name": "Alice", "age": 30, "city": "NYC"}',
};

async function echoAgent(input: unknown): Promise<string> {
  await new Promise((r) => setTimeout(r, 5)); // simulate latency
  return RESPONSES[String(input)] ?? "I don't know.";
}

// ---------------------------------------------------------------------------
// Test suite
// ---------------------------------------------------------------------------

function makeTestSuite(): TestCase[] {
  const containsDigit = (_expected: unknown, actual: unknown): number =>
    /\d/.test(String(actual)) ? 1.0 : 0.0;

  return [
    new TestCase({
      id: "math-basic",
      description: "Simple arithmetic",
      input: "What is 2 + 2?",
      expectedOutput: "4",
      scoring: exactMatch(),
      tags: ["math", "basic"],
    }),
    new TestCase({
      id: "geography-capital",
      description: "Capital city lookup",
      input: "What is the capital of France?",
      expectedOutput: "Paris",
      scoring: containsMatch(),
      tags: ["geography"],
    }),
    new TestCase({
      id: "greeting-fuzzy",
      description: "Greeting should be warm",
      input: "Say hello",
      expectedOutput: "Hello! How can I help?",
      scoring: fuzzyMatch(0.65),
      tags: ["greeting"],
    }),
    new TestCase({
      id: "json-output",
      description: "Agent returns valid JSON with expected fields",
      input: "Return JSON",
      expectedOutput: { name: "Alice", age: 30 },
      scoring: jsonMatch(["city"], 1.0),
      tags: ["structured"],
    }),
    new TestCase({
      id: "custom-number-check",
      description: "Response must contain a number",
      input: "What is 2 + 2?",
      expectedOutput: "4",
      scoring: customScorer(containsDigit, 1.0),
      tags: ["math", "custom"],
    }),
    new TestCase({
      id: "unknown-input",
      description: "Agent gracefully handles unknown queries",
      input: "xyzzy-unknown-prompt",
      expectedOutput: "I don't know.",
      scoring: exactMatch(),
      tags: ["edge-case"],
    }),
  ];
}

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------

async function main() {
  const harness = new EvalHarness({
    agent: echoAgent,
    suiteName: "basic-echo-suite",
    defaultTimeoutMs: 5_000,
    verbose: true,
  });

  harness.addMany(makeTestSuite());

  const result = await harness.run({
    reportJson: "reports/basic_report.json",
    reportHtml: "reports/basic_report.html",
  });

  console.log(
    `Programmatic: pass_rate=${(result.results.filter((r) => r.passed).length / result.results.length * 100).toFixed(1)}%`
  );

  process.exit(result.results.every((r) => r.passed) ? 0 : 1);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
